helper/schema/shims.go

```diff
@@ -94,3 +88,46 @@ func JSONMapToStateValue(m map[string]interface{}, block *configschema.Block) (c
 func StateValueFromInstanceState(is *terraform.InstanceState, ty cty.Type) (cty.Value, error) {
 	return is.AttrsAsObjectValue(ty)
 }
+
+// LegacyResourceSchema takes a *Resource and returns a deep copy with 0.12 specific
+// features removed. This is used by the shims to get a configschema that
+// directly matches the structure of the schema.Resource.
+func LegacyResourceSchema(r *Resource) *Resource {
+	if r == nil {
+		return nil
+	}
+	// start with a shallow copy
+	newResource := new(Resource)
+	*newResource = *r
+	newResource.Schema = map[string]*Schema{}
+
+	for k, s := range r.Schema {
+		newResource.Schema[k] = LegacySchema(s)
+	}
+
+	return newResource
+}
+
+// LegacySchema takes a *Schema and returns a deep copy with 0.12 specific
+// features removed. This is used by the shims to get a configschema that
+// directly matches the structure of the schema.Resource.
+func LegacySchema(s *Schema) *Schema {
+	if s == nil {
+		return nil
+	}
+	// start with a shallow copy
+	newSchema := new(Schema)
+	*newSchema = *s
+	newSchema.ConfigMode = SchemaConfigModeAuto
+	newSchema.PromoteSingle = false
```

Comment:
This one is actually a _pre-0.12-only_ feature, so I think we should probably leave it alone here. If any provider had set it then forcing it to `false` could be a breaking change for existing configurations when running in 0.11 mode.

(There's no special code handling `PromoteSingle` in v0.12 mode because it's actually the validation in Terraform Core that "breaks" this in v0.12, not anything the SDK does... in v0.12 the single-value doesn't get a chance to be promoted because Terraform Core rejects the single value during decoding, before the provider even gets to see it.)

Reply:
Good catch. I'll just automatically added all new flags.